File: ruby/component/timer.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the same terms as Ruby.

class Timer < OSX::NSObject
  include OSX
  attr_accessor :delegate
  
  def active?
    @timer != nil
  end
  
  def start(interval)
    stop if active?
    @timer = NSTimer.scheduledTimerWithTimeInterval_target_selector_userInfo_repeats(interval, self, 'onTimer:', nil, true)
    NSRunLoop.currentRunLoop.addTimer_forMode(@timer, NSEventTrackingRunLoopMode)
  end
  
  def stop
    return unless active?
    @timer.invalidate
    @timer = nil
  end
  
  def onTimer(sender)
    if active? && @delegate
      @delegate.timer_onTimer(self)
    end
  end
end

File: objc/Library/Timer.m
// Created by Satoshi Nakagawa.
// You can redistribute it and/or modify it under the Ruby's license or the GPL2.

#import "Timer.h"


@implementation Timer

@synthesize delegate;
@synthesize selector;

- (id)init
{
	if (self = [super init]) {
		selector = @selector(timerOnTimer:);
	}
	return self;
}

- (void)dealloc
{
	[self stop];
	[super dealloc];
}

- (BOOL)isActive
{
	return timer != nil;
}

- (void)start:(int)interval
{
	[self stop];
	
	timer = [[NSTimer scheduledTimerWithTimeInterval:interval target:self selector:@selector(onTimer:) userInfo:nil repeats:YES] retain];
	[[NSRunLoop currentRunLoop] addTimer:timer forMode:NSEventTrackingRunLoopMode];
}

- (void)stop
{
	[timer invalidate];
	[timer release];
	timer = nil;
}

- (void)onTimer:(id)sender
{
	if (!self.isActive) return;
	
	if ([delegate respondsToSelector:selector]) {
		[delegate performSelector:selector withObject:self];
	}
}

@end
